### ISSUE TITLE
Implement canonical reference counting plan and RCImmix

### ISSUE THREAD

**Issue by wks** [mmtk/mmtk-core#783]
TL;DR: We need to port the RC and RCImmix plans from JikesRVM.

# The algorithms

Reference counting is one of the basic canonical GC algorithms alongside MarkSweep, MarkCompact, and Semispace.  The notable difference is its use of reference counting instead of tracing as the way to identify garbage.  We need to include RC as well for completeness.

Here we list the algorithms we need to port to the Rust version of mmtk-core, and their characteristics:

-   RC
    -   Based on the RC plan in JikesRVM MMTk.
        -   Use the same free-list allocator as MarkSweep
        -   Primarily use RC, but also use back-up tracing.
        -   Use deferral
        -   Use coalescing
    -   Apply optimisations introduced by [Shahriyar et al. in ISMM'12](https://users.cecs.anu.edu.au/~steveb/pubs/papers/rc-ismm-2012.pdf)
        -   Limit the reference count bits.
        -   Lazy mod-buf insertion.
        -   Objects are allocated dead.
-   RCImmix ([Shahriyar et al. in OOPSLA'13](https://users.cecs.anu.edu.au/~steveb/pubs/papers/rcix-oopsla-2013.pdf))
    -   Use the Immix allocator
    -   Maintain a live-object counter for each Immix line
    -   Move objects (1) when a young object survived a GC, and (2) when doing full-heap backup tracing

# Related work

The [LXR GC algorithm](https://users.cecs.anu.edu.au/~steveb/pubs/papers/lxr-pldi-2022.pdf) is based on RCImmix.  The code is [in a separate branch](https://github.com/wenyuzhao/mmtk-core/commits/lxr), but is not merged yet.

**Comment by k-sareen** [comment]
From memory there are/were flags that you could turn on and off to approximate RCImmix in LXR itself. In fact I belive Table 7 in the LXR paper demonstrates this. @wenyuzhao can clarify further.
